File: frontend/services/catalog.js
import request, {getBaseUrl, getFullUrl} from "../lib/request"
import getFlag from "../lib/getFlag";

export const itemNameToEncodedName = (str) => {
    if (typeof str !== 'string') {
        str = '';
    }
    // https://stackoverflow.com/questions/987105/asp-net-mvc-routing-vs-reserved-filenames-in-windows
    return str.replace(/'/g, "")
        .replace(/[^a-zA-Z0-9]+/g, "-")
        .replace(/^-+|-+$/g, "")
        .replace(/^(COM\d|LPT\d|AUX|PRT|NUL|CON|BIN)$/i, "") || "unnamed";
}

const itemPageLate2016Enabled = getFlag('itemPageLate2016Enabled', false);
const csrEnabled = getFlag('clientSideRenderingEnabled', false);

export const getItemUrl = ({ assetId, name }) => {
    return `/catalog/${assetId}/${itemNameToEncodedName(name)}`;
}

export const searchCatalog = ({
                                  category,
                                  subCategory = null,
                                  query = null,
                                  limit,
                                  cursor = null,
                                  sort,
                                  creatorType = null,
                                  creatorId = null,
                                  creatorName = null,
                                  includeNotForSale = null
                              }) => {
    let url = '/v1/search/items?category=' + category + '&limit=' + limit + '&sortType=' + sort;
    if (cursor) {
        url += '&cursor=' + encodeURIComponent(cursor);
    }
    if (query) {
        url += '&keyword=' + encodeURIComponent(query);
    }
    if (subCategory) {
        url += '&subcategory=' + encodeURIComponent(subCategory);
    }
    if (creatorType && creatorId) {
        url += '&creatorTargetId=' + creatorId + '&creatorType=' + creatorType;
    }
    if (!creatorId && creatorName) {
        url += '&creatorName=' + creatorName + '&creatorType=1';
    }
    if (includeNotForSale === true) {
        url += '&includeNotForSale=true';
    }
    return request('GET', getFullUrl('catalog', url)).then(d => d.data);
}

/**
 * Only use this on server-side requests.
 * @param {number} assetId
 * @returns ProductInfoLegacy
 */
export const getProductInfoLegacy = async (assetId) => {
    return request('GET', getFullUrl('api', '/marketplace/productinfo?assetId=' + assetId)).then(d => d.data);
}

/**
 * @typedef {Object} ProductInfoLegacy
 * @property {number} TargetId - Unique identifier for the target of the product (typically the game or asset).
 * @property {string} ProductType - The type of product, e.g., "Game Pass".
 * @property {number} AssetId - ID of the associated asset, if any (0 if not applicable).
 * @property {number} ProductId - Unique identifier for the product.
 * @property {string} Name - Name of the game pass.
 * @property {string} Description - Description of the game pass.
 * @property {number} AssetTypeId - The type of asset (0 if not applicable).
 * @property {Object} Creator - Information about the creator of the game pass.
 * @property {number} Creator.Id - ID of the creator.
 * @property {string} Creator.Name - Username of the creator.
 * @property {string} Creator.CreatorType - Type of creator (e.g., "User").
 * @property {number} Creator.CreatorTargetId - ID used for referencing the creator.
 * @property {number} IconImageAssetId - ID of the icon image used for the game pass.
 * @property {string} Created - ISO 8601 timestamp of when the game pass was created.
 * @property {string} Updated - ISO 8601 timestamp of when the game pass was last updated.
 * @property {number} PriceInRobux - Price of the game pass in Robux.
 * @property {?number} PriceInTickets - Deprecated. Always null (Roblox tickets no longer exist).
 * @property {number} Sales - Number of times the game pass has been sold.
 * @property {boolean} IsNew - Indicates if the game pass is newly created.
 * @property {boolean} IsForSale - Whether the game pass is currently for sale.
 * @property {boolean} IsPublicDomain - Whether the asset is public domain (usually false).
 * @property {boolean} IsLimited - Whether the item is limited.
 * @property {boolean} IsLimitedUnique - Whether the item is a limited unique.
 * @property {number} Remaining - Number of items remaining (usually 0 for non-limiteds).
 * @property {number} MinimumMembershipLevel - Minimum membership level required to buy (0 = none).
 * @property {number} ContentRatingTypeId - Content rating (0 = all audiences).
 */

export const getItemDetails = async (assetIdArray) => {
    if (assetIdArray.length === 0) return { data: { data: [] } }
    while (true) {
        try {
            const res = await request('POST', getFullUrl('catalog', '/v1/catalog/items/details'), {
                items: assetIdArray.map(v => {
                    return {
                        itemType: 'Asset',
                        id: v,
                    }
                })
            });
            for (const item of res.data.data) {
                if (typeof item.isForSale === 'undefined') {
                    item.isForSale = (item.unitsAvailableForConsumption !== 0 && typeof item.price === 'number' && typeof item.lowestPrice === 'undefined');
                }
            }
            return res;
        } catch (e) {
            // @ts-ignore
            if (e.response && e.response.status === 429 && process.browser) {
                await new Promise((res) => setTimeout(res, 2500));
                continue;
            }
            throw e;
        }
    }
}

/**
 * @param {number[]} assetIdArray
 * @returns {Promise<AveriaCollection<AssetDetailsEntry>>}
 */
export const getItemDetailsNew = async (assetIdArray) => {
    if (assetIdArray.length === 0) return Promise.resolve({ data: [] });
    
    try {
        /** @type RawAssetDetails[] */
        const response = (await request("POST", getFullUrl("catalog", "/v1/catalog/items/details"), {
            items: assetIdArray.map(v => {
                return {
                    itemType: 'Asset',
                    id: v,
                }
            })
        })).data.data;
        
        for (const item of response) {
            if (typeof item.isForSale === 'undefined') {
                item.isForSale = (item.unitsAvailableForConsumption !== 0 && typeof item.price === 'number' && typeof item.lowestPrice === 'undefined');
            }
        }
        
        return response;
    } catch (e) {
        // @ts-ignore
        if (e.response && e.response.status === 429 && process.browser) {
            await new Promise((res) => setTimeout(res, 2500));
            return getItemDetailsNew(assetIdArray);
        }
        throw e;
    }
}

/**
 * @param {number} assetId
 * @param {number} assetTypeId
 * @param {number} limit
 * @returns {Promise<AveriaCollection<RecommendedItemEntry>>}
 */
export const getRecommendations = ({ assetId, assetTypeId, limit }) => {
    return request('GET', getFullUrl('catalog', '/v1/recommendations/asset/' + assetTypeId + '?contextAssetId=' + assetId + '&numItems=' + limit)).then(d => d.data);
}

export const getComments = async ({ assetId, offset }) => {
    return request('GET', getBaseUrl() + '/comments/get-json?assetId=' + assetId + '&startIndex=' + offset + '&thumbnailWidth=100&thumbnailHeight=100&thumbnailFormat=PNG&cachebuster=' + Math.random()).then(d => d.data);
}

export const createComment = async ({ assetId, comment }) => {
    let result = await request('POST', getBaseUrl() + '/comments/post', {
        text: comment,
        assetId: assetId,
    });
    if (typeof result.data.ErrorCode === 'string') {
        throw new Error(result.data.ErrorCode);
    }
    return result.data;
}

export const addOrRemoveFromCollections = ({ assetId, addToProfile }) => {
    return request('POST', getBaseUrl() + '/asset/toggle-profile', {
        assetId,
        addToProfile,
    })
}

export const getIsFavorited = async ({ assetId, userId }) => {
    return await request('GET', getFullUrl('catalog', '/v1/favorites/users/' + userId + '/assets/' + assetId + '/favorite')).then(d => d.data);
}

export const createFavorite = async ({ assetId, userId }) => {
    return await request('POST', getFullUrl('catalog', '/v1/favorites/users/' + userId + '/assets/' + assetId + '/favorite'));
}

export const deleteFavorite = async ({ assetId, userId }) => {
    return await request('DELETE', getFullUrl('catalog', '/v1/favorites/users/' + userId + '/assets/' + assetId + '/favorite'));
}

export const getAudio = async ({ audioId }) => {
    return await request('GET', `${getBaseUrl()}/asset/?id=${audioId}`).then(d => d.data);
}

export const getAudioURL = async ({ audioId }) => {
    return `${getBaseUrl()}/asset/?id=${audioId}`;
}

export const getModerationStatus = ({ assetID }) => {
    return request('GET', `${getBaseUrl()}/asset/status?assetId=${assetID}`).then(d => d?.data?.moderationStatus);
}

/**
 * @typedef {Object} RawAssetDetails
 * @property {number} id - Unique identifier for the item.
 * @property {number} assetType - Type ID of the asset.
 * @property {string} name - Name of the item.
 * @property {string} description - Description of the item.
 * @property {string[]} genres - Array of genres the item belongs to.
 * @property {"User" | "Group"} creatorType - Type of creator.
 * @property {number} creatorTargetId - ID of the user or group who created the item.
 * @property {string} creatorName - Name of the creator.
 * @property {string|null} offsaleDeadline - When the item will go off sale, if applicable.
 * @property {("Limited" | "LimitedUnique")[]} itemRestrictions - Any special item restrictions.
 * @property {number} saleCount - Total number of sales.
 * @property {"Asset" | string} itemType - Type of the item.
 * @property {number} favoriteCount - Number of times the item was favorited.
 * @property {boolean} isForSale - Whether the item is currently for sale.
 * @property {boolean} commentsEnabled - Whether comments are enabled.
 * @property {number|null} price - Original sale price in Robux.
 * @property {number|null} priceTickets - Original sale price in Tickets (legacy).
 * @property {number|null} lowestPrice - Current lowest resale price.
 * @property {null|string} priceStatus - Status of the price, if applicable.
 * @property {Object|null} lowestSellerData - Information about the current lowest seller.
 * @property {number} lowestSellerData.userId - ID of the seller.
 * @property {string} lowestSellerData.username - Username of the seller.
 * @property {number} lowestSellerData.userAssetId - ID of the specific user-owned asset.
 * @property {number} lowestSellerData.price - Resale price set by the seller.
 * @property {number} lowestSellerData.assetId - ID of the asset being sold.
 * @property {number|null} unitsAvailableForConsumption - Units available (for items like gear or use-once assets).
 * @property {number} serialCount - Total number of serialized copies.
 * @property {boolean} is18Plus - Whether the item is age-restricted.
 * @property {"ReviewApproved" | string} moderationStatus - Moderation status of the item.
 * @property {string} createdAt - ISO timestamp of when the item was created.
 * @property {string} updatedAt - ISO timestamp of the last update.
 */

/**
 * @typedef {RawAssetDetails & {
 *   state: string;
 *   imageUrl: string;
 *   owned: boolean;
 * }} CatalogAssetDetails
 */

/**
 * @typedef CatalogResultsMetadata
 * @property {number} limit
 * @property {number} page
 * @property {string|null} cursor
 * @property {string|null} nextCursor
 * @property {string|null} prevCursor
 */

/**
 * @param {number} category
 * @param {number|null} subCategory
 * @param {string|null} query
 * @param {number[]|null} genres
 * @param {number} limit
 * @param {string|null} cursor
 * @param {number} sort
 * @param {number|null} creatorType
 * @param {number|null} creatorId
 * @param {string|null} creatorName
 * @param {boolean|null} includeNotForSale
 * @param {number|null} priceOption
 * @param {[number, number]|null} priceRange
 * @param {number|null} currency
 * @returns {Promise<AveriaCollectionPaginated<{itemType: string; id: number;}>>}
 */
export const searchCatalog2 = ({
                                   category,
                                   subCategory = null,
                                   query = null,
                                   genres = null,
                                   limit,
                                   cursor = null,
                                   sort,
                                   creatorType = null,
                                   creatorId = null,
                                   creatorName = null,
                                   includeNotForSale = false,
                                   priceOption = null,
                                   priceRange = null,
                                   currency = null,
                               }) => {
    let url = '/v3/search/items?category=' + category + '&limit=' + limit + '&sortType=' + sort;
    if (cursor) {
        url += '&cursor=' + encodeURIComponent(cursor);
    }
    if (query) {
        url += '&keyword=' + encodeURIComponent(query);
    }
    if (subCategory) {
        url += '&subcategory=' + encodeURIComponent(subCategory);
    }
    if (genres) {
        url += `&genres=${genres.join(",")}`;
    }
    if (creatorType) {
        url += '&creatorType=' + encodeURIComponent(creatorType);
    }
    if (creatorType && creatorId) {
        url += '&creatorTargetId=' + encodeURIComponent(creatorId);
    }
    if (!creatorId && creatorName) {
        url += '&creatorName=' + encodeURIComponent(creatorName) + '&creatorType=1';
    }
    if (includeNotForSale === true) {
        url += '&includeNotForSale=true';
    }
    if (priceOption) {
        url += '&priceOption=' + priceOption;
    }
    if (priceRange) {
        url += `&minPrice=${priceRange[0]}&maxPrice=${priceRange[1]}`;
    }
    if (currency) {
        url += '&currency=' + currency;
    }
    return request('GET', getFullUrl('catalog', url)).then(d => d.data);
}

/**
 * @param {{id: number; itemType: string;}[]} assetIdArray
 * @returns {Promise<AssetDetailsEntry[]|null>}
 */
export const getAssetDetailsClean = (assetIdArray) => {
    return request("POST", getFullUrl("catalog", "/v1/catalog/items/details"), {
        items: assetIdArray,
    }).then(d => d?.data?.data);
}

/**
 * @returns {Promise<CatalogNavigationClass>}
 */
export const getNavigationMenuItems = () => {
    return request('GET', getFullUrl("catalog", "/v3/navigation-menu-items")).then(d => d?.data);
}

/**
 * @typedef {Object} CatalogSubCategory
 * @property {string} subCategory
 * @property {number} subCategoryId
 * @property {string|null} name
 * @property {number[]} assetTypeIds
 */

/**
 * @typedef {Object} CatalogCategory
 * @property {string} category
 * @property {number} categoryId
 * @property {string} name
 * @property {number} orderIndex
 * @property {CatalogSubCategory[]} subCategories
 * @property {number[]} assetTypeIds
 */

/**
 * @typedef {Object} GenreClass
 * @property {number} genreId
 * @property {string|undefined|null} name
 * @property {string} genre
 */

/**
 * @typedef {Object} CatalogNavigationClass
 * @property {CatalogCategory[]} categories
 * @property {GenreClass[]} genres
 */

File: Roblox/Roblox.Dto/Assets/Catalog.cs
using Roblox.Models;
using Roblox.Models.Assets;
using Roblox.Models.Economy;
using Type = Roblox.Models.Assets.Type;

namespace Roblox.Dto.Assets;

public class CatalogSearchRequest
{
    public string? category { get; set; }
    public string? subcategory { get; set; }
    public string? sortType { get; set; }
    public string? keyword { get; set; }
    public string? cursor { get; set; }
    public int limit { get; set; } = 10;
    public CreatorType? creatorType { get; set; }
    public long? creatorTargetId { get; set; }
    public bool includeNotForSale { get; set; } = false;
    public IEnumerable<Genre>? genres { get; set; }
    public bool include18Plus { get; set; }
}

public class CatalogMultiGetEntry
{
    public string itemType { get; set; }
    public long id { get; set; }

    [System.Text.Json.Serialization.JsonIgnore]
    public bool isOnSale { get; set; }
}

public class SearchResponse : RobloxCollectionPaginated<CatalogMultiGetEntry>
{
    public SearchResponse()
    {
        data = Array.Empty<CatalogMultiGetEntry>();
    }
    // es extension
    public int _total { get; set; }
    public string? keyword { get; set; }
    public object elasticsearchDebugInfo { get; set; }
}

public class ItemRestrictions {
    public long assetId { get; set; }
    public bool isLimited { get; set; }
    public bool isLimitedUnique { get; set; }
    public bool exists { get; set; } = true;
}

public class MinimalCatalogEntry
{
    public long assetId { get; set; }
    public long? priceRobux { get; set; }
    public long? priceTickets { get; set; }
    public bool isForSale { get; set; }
    public bool isLimited { get; set; }
    public bool isLimitedUnique { get; set; }
    // Current sale total
    public long saleCount { get; set; }
    // Amount of serials available for purchase
    public long serialCount { get; set; }
    public CreatorType creatorType { get; set; }
    public long creatorId { get; set; }
    public DateTime? offsaleAt { get; set; }
    public Type assetType { get; set; }
    public bool IsFree()
    {
        return isForSale && priceRobux == 0 && priceTickets == null;
    }
}

public class UserAssetForSaleEntry
{
    public long userAssetId { get; set; }
    public long userId { get; set; }
    public string username { get; set; }
    public int? serialNumber { get; set; }
    public long price { get; set; }
    public long assetId { get; set; }
}

public class AssetResaleChartEntry
{
    public long value { get; set; }
    public DateTime date { get; set; }
}

public class RecentAveragePrice
{
    public long? recentAveragePrice { get; set; }
}

public class AssetResaleCharts
{
    public IEnumerable<AssetResaleChartEntry> priceDataPoints { get; set; }
    public IEnumerable<AssetResaleChartEntry> volumeDataPoints { get; set; }
}

public class AssetResaleData
{
    public long assetStock { get; set; }
    public long sales { get; set; }
    public long numberRemaining { get; set; }
    public long recentAveragePrice { get; set; }
    public IEnumerable<AssetResaleChartEntry> priceDataPoints { get; set; }
    public IEnumerable<AssetResaleChartEntry> volumeDataPoints { get; set; }
}

public class CatalogSearchRequest2
{
    public int? category { get; set; }
    public int? subcategory { get; set; }
    public int? sortType { get; set; }
    public string? keyword { get; set; }
    public string? cursor { get; set; }
    public int limit { get; set; } = 10;
    public CreatorType? creatorType { get; set; }
    public long? creatorTargetId { get; set; }
    public bool includeNotForSale { get; set; } = false;
    public IEnumerable<Genre>? genres { get; set; }
    public PriceOption? priceOption { get; set; }
    public CurrencyType2? currency { get; set; }
    public long? minPrice { get; set; }
    public long? maxPrice { get; set; }
}

public enum PriceOption
{
    Any = 0,
    Range,
    Free,
}

public static class CatalogCategory
{
    public const int Featured = 0;
    public const int All = 1;
    public const int Collectibles = 2;
    public const int Accessories = 11;
    public const int Clothing = 3;
    public const int BodyParts = 4;
    public const int Gears = 5;
    public const int AvatarAnimations = 12;
    public const int Emotes = 13;
    public const int Special = 69;
}

public static class CatalogSubCategory
{
    public const int All = 0;
    public const int Faces = 10;
    public const int Packages = 37;
    public const int Shirts = 12;
    public const int TeeShirts = 13;
    public const int Pants = 14;
    // GEAR
    public const int Building = 8;
    public const int Explosive = 3;
    public const int Melee = 1;
    public const int Musical = 6;
    public const int Navigation = 5;
    public const int Powerup = 4;
    public const int Ranged = 2;
    public const int Social = 7;
    public const int Transport = 9;
    //
    public const int Accessories = 19;
    public const int Animations = 39;
    public const int Gear = 5;
    public const int Emotes = 39;
    public const int Heads = 15;
    public const int FaceAccessories = 27;
    public const int Hats = 9;
    public const int Hair = 20;
    public const int Neck = 22;
    public const int Shoulder = 23;
    public const int Front = 24;
    public const int Back = 25;
    public const int Waist = 26;

    public const int Owned = 101;
    public const int Wearing = 102;
}

public class CatalogSubCategoryData
{
    public string subCategory { get; set; }
    public int subCategoryId { get; set; }
    public string? name { get; set; }
    public List<long> assetTypeIds { get; set; }
}

public class CatalogCategoryData
{
    public string category { get; set; }
    public int categoryId { get; set; }
    public string name { get; set; }
    public int orderIndex { get; set; }
    public List<CatalogSubCategoryData> subCategories { get; set; }
    public List<long> assetTypeIds { get; set; }
}

public class GenreClass
{
    public int genreId { get; set; }
    public string? name { get; set; }
    public string genre { get; set; }
}

public class CatalogNavigationClass
{
    public List<CatalogCategoryData> categories { get; set; }
    public List<GenreClass> genres { get; set; }
}
